### PR TITLE
Source MSSQL: add option to disable Snapshot mode and initial backup

### PR DESCRIPTION
**Pull request by sivankumar86**
## What
*Describe what the change is solving*
   Added option to disable initial snapshot and enabling snapshot mode as we use read_committed replica which is not required to enable snapshot isolation 

https://github.com/airbytehq/airbyte/issues/12592

## How
   There is a option in debezium connector to disable snapshot isolation mode. 

     // below option would not lock the table 
      
    props.setProperty("snapshot.isolation.mode", "read_committed");
   
    //below. option would not take initial snapshot and only capture changes 
    
    props.setProperty("snapshot.mode", "schema_only");

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Comment by sivankumar86** [comment]
> [CLA assistant check](https://cla-assistant.io/airbytehq/airbyte?pullRequest=12759) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=12759) before we can accept your contribution.
> 
> **Sivakumar Ramaswamy** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=12759) it.

i did digital sign but, it is not working

**Comment by alafanechere** [comment]
@sivankumar86 thank you for this contribution! To fix the CLA signature I think you need to make sure that your git config email is the same as your using for GitHub. You committed with your Gmail address, is it the same email as your github account email?

**Comment by sivankumar86** [comment]
> @sivankumar86 thank you for this contribution! To fix the CLA signature I think you need to make sure that your git config email is the same as your using for GitHub. You committed with your Gmail address, is it the same email as your github account email?

Thanks for reply. I added email address which was used to commit and it got resolved.

**Comment by alafanechere** [comment]
I'd love a review from the connector team on this. @sivankumar86 I'm wondering to what extent this change can be propagated to other CDC compatible source databases. Is anything specific to MSSQL in your modifications?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java [hunk 1]
```diff
@@ -408,6 +409,29 @@ private static AirbyteStream addCdcMetadataColumns(final AirbyteStream stream) {
     return stream;
   }
 
+  //validate snapshot isolation and cdc mode
+  private void addCDCDetails(final JsonNode cdcMethod,Properties props) {
```
Comment (tuliren) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:413]: The logics here can be merged into the `getDebeziumProperties` and unit tested.

**Review by tuliren — CHANGES_REQUESTED** [review]
@sivankumar86, thank you for the contribution!

The overall design looks good. However, since multiple configurations have been introduced to the CDC method, the `replication_method` needs to be changed to an object. In addition, I think some unit tests and a new test case in the acceptance test are necessary.

I will create a PR based on your PR, and make all the changes mentioned in the review. When all those are done, I will merge your PR and then my patch.

**Comment by subodh1810** [comment]
Should this change be across Postgres and Mysql connectors as well for consistency?

**Comment by tuliren** [comment]
> Should this change be across Postgres and Mysql connectors as well for consistency?

Yes.

File: airbyte-integrations/connectors/source-mssql/src/main/resources/spec.json [hunk 2]
```diff
@@ -111,6 +111,20 @@
         "default": "STANDARD",
         "enum": ["STANDARD", "CDC"],
         "order": 8
+      },
+      "is_snapshotDisabled": {
```
Comment (grishick) [inline airbyte-integrations/connectors/source-mssql/src/main/resources/spec.json:115]: Let's not mix camel case and snake case in the same identifier. Property names have so far been following snake case, so let's stick with that

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java [hunk 3]
```diff
@@ -408,6 +409,29 @@ private static AirbyteStream addCdcMetadataColumns(final AirbyteStream stream) {
...
+      // https://docs.microsoft.com/en-us/sql/t-sql/statements/set-transaction-isolation-level-transact-sql?view=sql-server-ver15
+      // https://debezium.io/documentation/reference/1.4/connectors/sqlserver.html#sqlserver-property-snapshot-isolation-mode
+      // we set this to avoid preventing other (non-Airbyte) transactions from updating table rows while
+      // we snapshot
+      props.setProperty("snapshot.isolation.mode", "snapshot");
+    }
+    if(cdcMethod.hasNonNull("is_cdc_only") &&
+            cdcMethod.get("is_cdc_only").asBoolean()){
+      props.setProperty("snapshot.mode", "schema_only");
+    }else{
```
Comment (grishick) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:428]: nit: add space after '}'

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java [hunk 4]
```diff
@@ -408,6 +409,29 @@ private static AirbyteStream addCdcMetadataColumns(final AirbyteStream stream) {
     return stream;
   }
 
+  //validate snapshot isolation and cdc mode
+  private void addCDCDetails(final JsonNode cdcMethod,Properties props) {
+    if(cdcMethod.hasNonNull("is_snapshotDisabled") &&
+            cdcMethod.get("is_snapshotDisabled").asBoolean()) {
+      props.setProperty("snapshot.isolation.mode", "read_committed");
+    }
```
Comment (grishick) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:417]: nit: preferred formatting:
'} else {'

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java [hunk 5]
```diff
@@ -408,6 +409,29 @@ private static AirbyteStream addCdcMetadataColumns(final AirbyteStream stream) {
...
+    else{
+      // https://docs.microsoft.com/en-us/sql/t-sql/statements/set-transaction-isolation-level-transact-sql?view=sql-server-ver15
+      // https://debezium.io/documentation/reference/1.4/connectors/sqlserver.html#sqlserver-property-snapshot-isolation-mode
+      // we set this to avoid preventing other (non-Airbyte) transactions from updating table rows while
+      // we snapshot
+      props.setProperty("snapshot.isolation.mode", "snapshot");
+    }
+    if(cdcMethod.hasNonNull("is_cdc_only") &&
+            cdcMethod.get("is_cdc_only").asBoolean()){
```
Comment (grishick) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:426]: nit: space before '{'

**Comment by tuliren** [comment]
> Should this change be across Postgres and Mysql connectors as well for consistency?

I looked into this more closely. Different databases actually have different CDC settings. Since this is not a planned sprint project, I will not apply this change to Postgres or MySQL this time. Will create a ticket to track it in the future.

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlCdcProperties.java [hunk 6]
```diff
@@ -29,7 +20,7 @@ static Properties getDebeziumProperties() {
     props.setProperty("converters", "mssql_converter");
     props.setProperty("mssql_converter.type", "io.airbyte.integrations.debezium.internals.MSSQLConverter");
 
-    return props;
+    return props;
```
[the remark this answers is not in the document]
Comment (sivankumar86) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlCdcProperties.java]: removed extra space

File: airbyte-integrations/connectors/source-mssql/src/main/resources/spec.json (hunk 2 above)
Comment (sivankumar86) [reply airbyte-integrations/connectors/source-mssql/src/main/resources/spec.json]: followed only snake case.

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java (hunk 3 above)
Comment (sivankumar86) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java]: followed same standard.

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java (hunk 5 above)
Comment (sivankumar86) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java]: cleaned up

File: docs/integrations/sources/mssql.md [hunk 7]
```diff
@@ -63,7 +63,7 @@ Please read the [CDC docs](../../understanding-airbyte/cdc.md) for an overview o
 * There are some critical issues regarding certain datatypes. Please find detailed info in [this Github issue](https://github.com/airbytehq/airbyte/issues/4542).
 * CDC is only available for SQL Server 2016 Service Pack 1 \(SP1\) and later.
 * _db\_owner_ \(or higher\) permissions are required to perform the [neccessary setup](mssql.md#setting-up-cdc-for-mssql) for CDC.
-* You must enable [snapshot isolation mode](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql/snapshot-isolation-in-sql-server) on the database\(s\) you want to sync. This is used for retrieving an initial snapshot without locking tables.
+* You must enable [snapshot isolation mode](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql/snapshot-isolation-in-sql-server) on the database\(s\) you want to sync. This is used for retrieving an initial snapshot without locking tables. You can disable  snapshot isolation and initial load in connector setting if required.
```
[the remark this answers is not in the document]
Comment (sivankumar86) [reply docs/integrations/sources/mssql.md]: done

**Comment by sivankumar86** [comment]
@tuliren  Thank you for providing review/feedback. I have changed the code and it is specific to mssql server hence, it may not suitable for other connector.  could you take a look on new changes ?

**Comment by alafanechere** [comment]
I handover the assignment to @grishick as the DB connector team is currently working on this.

**Comment by tuliren** [comment]
@alafanechere, @grishick, I am already working on this. Please let me push this to the finishing line. Switching the assignee at this moment will slow it down.

@sivankumar86, thank you very much. You can leave the PR as is. I am working on a PR on top of this to add a few more optimizations. Once this is one, I will merge this PR.

**Comment by sivankumar86** [comment]
> @alafanechere, @grishick, I am already working on this. Please let me push this to the finishing line. Switching the assignee at this moment will slow it down.
> 
> @sivankumar86, thank you very much. You can leave the PR as is. I am working on a PR on top of this to add a few more optimizations. Once this is one, I will merge this PR.

@[tuliren](https://github.com/tuliren) thank you for taking ownership. Could you please provide me your PR link so that I can subscribe and also, it would be great if you can push to main asap .

**Comment by tuliren** [comment]
@sivankumar86, the PR is here: https://github.com/airbytehq/airbyte/pull/13168

**Comment by tuliren** [comment]
I ran into some issue when trying to publish a new version for the MSSQL connector. Still working on it in https://github.com/airbytehq/airbyte/pull/13176.

**Pull request merged**